webdriver/tests/bidi/browsing_context/history_updated/history_updated.py

```diff
@@ -0,0 +1,216 @@
+import pytest
+
+from webdriver.bidi.modules.script import ContextTarget
+from webdriver.error import TimeoutException
+from tests.support.sync import AsyncPoll
+
+from ... import recursive_compare
+
+pytestmark = pytest.mark.asyncio
+
+EMPTY_PAGE = "/webdriver/tests/bidi/browsing_context/support/empty.html"
+FRAGMENT_NAVIGATED_EVENT = "browsingContext.fragmentNavigated"
+HISTORY_UPDATED_EVENT = "browsingContext.historyUpdated"
+CREATED_EVENT = "browsingContext.contextCreated"
+
+@pytest.mark.parametrize(
+    "hash_before, hash_after, history_method",
+    [
+        ("", "#foo", "pushState"),
+        ("#foo", "#bar", "pushState"),
+        ("#foo", "#foo", "pushState"),
+        ("#bar", "", "pushState"),
+        ("", "#foo", "replaceState"),
+        ("#foo", "#bar", "replaceState"),
+        ("#foo", "#foo", "replaceState"),
+        ("#bar", "", "replaceState"),
+    ]
+)
+async def test_history_url_update(
+    bidi_session, new_tab, url, subscribe_events, hash_before, hash_after, history_method
+):
+    target_context = new_tab["context"]
+
+    await bidi_session.browsing_context.navigate(
+        context=new_tab["context"], url=url(EMPTY_PAGE + hash_before), wait="complete"
+    )
+
+    await subscribe_events([FRAGMENT_NAVIGATED_EVENT, HISTORY_UPDATED_EVENT])
+
+    fragment_navigated_events = []
+    history_updated_events = []
+    async def on_event(method, data):
+        if method == FRAGMENT_NAVIGATED_EVENT:
+          fragment_navigated_events.append(data)
+        if method == HISTORY_UPDATED_EVENT:
+           history_updated_events.append(data)
+
+    remove_fragment_navigated_listener = bidi_session.add_event_listener(FRAGMENT_NAVIGATED_EVENT, on_event)
+    remove_history_updated_listener = bidi_session.add_event_listener(HISTORY_UPDATED_EVENT, on_event)
+
+    try:
+      target_url = url(EMPTY_PAGE + hash_after)
+
+      await bidi_session.script.call_function(
+          raw_result=True,
+          function_declaration="""(method, url) => {
+              history[method](null, null, url);
+          }""",
+          arguments=[
+              {"type": "string", "value": history_method},
+              {"type": "string", "value": target_url},
+          ],
```

Comment:
NIT: I guess this will be more readable
```suggestion
          function_declaration=f"""(url) => {{
              history.{history_method}(null, null, url);
          }}""",
          arguments=[
              {"type": "string", "value": target_url},
          ],
```